### ISSUE TITLE
Can't follow error in Problems tab in C++ project under WSL (re-open #2786)

### ISSUE THREAD

**Issue by dzvancuks** [microsoft/vscode-remote-release#2871]
Version: 1.44.0 (system setup)
Commit: 2aae1f26c72891c399f860409176fe435a154b13
Date: 2020-04-07T23:31:18.860Z
Electron: 7.1.11
Chrome: 78.0.3904.130
Node.js: 12.8.1
V8: 7.8.279.23-electron.0
OS: Windows_NT x64 10.0.18363
**WSL 1**

Steps to Reproduce:

1. Open pure makefile project in VSC under WSL, e.g. Ubuntu 16.04 with gcc and make.
2. Configure tasks.json for compilation (see below).
3. Create error. E.g.` int a = "a";`
4. Compile project using task.
5. Switch to PROBLEMS tab and click on error message.
5. Popup message will inform that:
> Unable to open 'file.cpp': Unable to read file 'vscode-remote://wsl+ubuntu-16.04/file.cpp' (Error: Unable to resolve non-existing file 'vscode-remote://wsl+ubuntu-16.04/file.cpp').

![image](https://user-images.githubusercontent.com/744090/79588426-95a4fe00-80dc-11ea-99a7-115005963cc0.png)

tasks.json:
 `{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Make project",
            "type": "shell",
            "command": "make",
            // use options.cwd property if the Makefile is not in the project root ${workspaceRoot} dir
            "options": {
                "cwd": "${workspaceRoot}"
            },
            // start the build without prompting for task selection, use "group": "build" otherwise
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "presentation": {
                "echo": true,
                "reveal": "always",
                "focus": false,
                "panel": "shared"
            },
            // arg passing example: in this case is executed make QUIET=0
            "args": [
                "QUIET=0"
            ],
            // Use the standard less compilation problem matcher.
            "problemMatcher": {
                "owner": "cpp",
                "fileLocation": [
                    "absolute"
                ],
                "pattern": {
                    "regexp": "^(.*):(\\d+):(\\d+):\\s+(warning|error):\\s+(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "severity": 4,
                    "message": 5
                }
            }
        }
    ]
}`

Same issue if change problemMatcher to `"fileLocation": "relative",`.

Attaching test project. Open it in WSL and hit Start Debugging in Run tab
[test.tar.gz](https://github.com/microsoft/vscode-remote-release/files/4521059/test.tar.gz)

Does this issue occur when you try this locally?: No
Does this issue occur when you try this locally and all extensions are disabled?: Without extension - Yes, this is not C++ extension issue. Locally - no, this is WSL1 issue.

**Comment by dzvancuks** [comment]
Re-opening issue https://github.com/microsoft/vscode-remote-release/issues/2786. Because it was closed without checking.

**Comment by lixin-wei** [comment]
I've met a similar error in remote mode while debugging using gdb. I'm using a centOS remote server.

It seems that current working path(CWD) is wrong. In my case `direct_actor_transport.cc` actually placed in `/home/admin/ray`(the opened project dir), but the error message shows vscode try to find it in `/home/admin/`(ssh home dir)

Can anyone have a look?

![image](https://user-images.githubusercontent.com/19600697/90486743-008a4280-e16c-11ea-9dce-e8aacb513ce4.png)

here is my launch.json

```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "GDB Attach",
            "type": "cppdbg",
            "request": "attach",
            "program": "/home/admin/py3/bin/python",
            "MIMode": "gdb",
            "processId": 81508,
            "setupCommands": [
                {
                    "description": "Enable pretty-printing for gdb",
                    "text": "-enable-pretty-printing",
                    "ignoreFailures": true
                }
            ],
            "miDebuggerPath": "/usr/bin/gdb",
        }
    ]
}
```

**Comment by dzvancuks** [comment]
@roblourens yes, path is correct

**Comment by roblourens** [comment]
I get something slightly different but there is an issue with your build task. You have `fileLocation` as "absolute" when the build output produces relative paths. So try this

```json
"fileLocation": [
    "relative",
    "${workspaceFolder}"
],
```


Alternately you should be able to replace the whole problemMatcher with the built-in one for gcc

```
"problemMatcher": [
    "$gcc"
]
```

then it actually works fine. Does that work for you?

**Comment by lixin-wei** [comment]
> @lixin-wei your issue probably belongs on the C++ extension repo. But where are you seeing `/home/admin`? It looks like the debugger thinks that file is in `/proc/self/cwd...`

When I click `Create File`, it creates a file in `/home/admin`, so `/proc/self/cwd` points to `/home/admin`.

Are you sure it's an issue of C++ extension but not remote development? Should I create an issue in  C++ extension repo?

**Comment by dzvancuks** [comment]
@roblourens, tried both solutions, unfortunately none of them worked

**Comment by roblourens** [comment]
> /proc/self/cwd points to /home/admin

I see, I read about what this special path actually does, and I don't think this assumption will hold up in VS Code. You can't assume that every process has your workspace folder as a CWD.

@dzvancuks did it change anything?

**Comment by lixin-wei** [comment]
@roblourens Ok, I report this issue in c++ extension's repo but has no response. Could you help me to cue someone that might be responsible for this problem?

**Comment by roblourens** [comment]
It's only been an hour and it's night. But the extension has on control over this. You just can't make assumptions about the cwd of vscode processes. What if you have multiple workspace folders in your process? We can't have all of them be the cwd. You should restructure it to not use that path.

**Comment by lixin-wei** [comment]
@roblourens OK, thanks, I'll wait for response.

> But the extension has on control over this. You just can't make assumptions about the cwd of vscode processes.

I think It's the extension that assumes /proc/self/cwd is the working dir and try to read file from it. I didn't set any configurations that use /proc/self/cwd.

**Comment by roblourens** [comment]
Oh, hm. I doubt that this comes from the extension but we can see if they have any ideas.

**Comment by dzvancuks** [comment]
> @dzvancuks did it change anything?
@roblourens, no error message was the same as in my screenshot above

**Comment by roblourens** [comment]
Ok. It still seems like there must be an issue with that path. I can't imagine what sort of problem would make vscode just fail to open a file with the correct path.

Is the casing correct? If you copy that file's path from vscode and compare it to the path in the error message, is it exactly character-by-character identical? Ignoring the remote name at the beginning.

**Comment by dzvancuks** [comment]
@roblourens , yes, path is all lower case. Works in console. My guess is that "remote name at the beginning" causes such issue

**Comment by roblourens** [comment]
@aeschli for WSL in case you have any idea. Sorry, I am stumped. I don't know where the full URI would be exposed so we can compare a working one to this error.

**Comment by michael-go** [comment]
I had a very similar issue and the solution here resolved it for me: https://github.com/microsoft/vscode-cpptools/issues/6019#issuecomment-680141032
(just need to fit the actual paths to your case)

**Comment by andrew-gillan** [comment]
I landed here trying to solve this issue myself. I was able to get things working by adding this to my tasks.json
`
"problemMatcher": {
    "base": "$gcc",
    "fileLocation": ["relative", "/workspaces/repo"]
}			},
`
For some reason it looks like VSCode has trouble following the paths with `remote://`, so the above manually replaces it with the correct path from inside the remote.

**Comment by cnwuwil** [comment]
when you try to compile with bazel, you will have 4 soft link(bazel-bin, bazel-{source-folder}, bazel-out, bazel-testlog)
modify this attribute in launch.json will be work

"cwd": "${workspaceFolder}/bazel-{source-folder}"
